pandas/_libs/meson.build

```diff
@@ -0,0 +1,165 @@
+    build_by_default: true,
+    build_always_stale: true
+)
+_algos_common_helper = custom_target('algos_common_helper_pxi',
+    output: 'algos_common_helper.pxi',
+    input: 'algos_common_helper.pxi.in',
+    command: [
+        py, tempita, '@INPUT@', '-o', '@OUTDIR@'
+    ]
+)
+_khash_primitive_helper = custom_target('khash_primitive_helper_pxi',
+    output: 'khash_for_primitive_helper.pxi',
+    input: 'khash_for_primitive_helper.pxi.in',
+    command: [
+        py, tempita, '@INPUT@', '-o', '@OUTDIR@'
+    ]
+)
+_hashtable_class_helper = custom_target('hashtable_class_helper_pxi',
+    output: 'hashtable_class_helper.pxi',
+    input: 'hashtable_class_helper.pxi.in',
+    command: [
+        py, tempita, '@INPUT@', '-o', '@OUTDIR@'
+    ]
+)
+_hashtable_func_helper = custom_target('hashtable_func_helper_pxi',
+    output: 'hashtable_func_helper.pxi',
+    input: 'hashtable_func_helper.pxi.in',
+    command: [
+        py, tempita, '@INPUT@', '-o', '@OUTDIR@'
+    ]
+)
+_index_class_helper = custom_target('index_class_helper_pxi',
+    output: 'index_class_helper.pxi',
+    input: 'index_class_helper.pxi.in',
+    command: [
+        py, tempita, '@INPUT@', '-o', '@OUTDIR@'
+    ]
+)
+_sparse_op_helper = custom_target('sparse_op_helper_pxi',
+    output: 'sparse_op_helper.pxi',
+    input: 'sparse_op_helper.pxi.in',
+    command: [
+        py, tempita, '@INPUT@', '-o', '@OUTDIR@'
+    ]
+)
+_intervaltree_helper = custom_target('intervaltree_helper_pxi',
+    output: 'intervaltree.pxi',
+    input: 'intervaltree.pxi.in',
+    command: [
+        py, tempita, '@INPUT@', '-o', '@OUTDIR@'
+    ]
+)
+_khash_primitive_helper_dep = declare_dependency(sources: _khash_primitive_helper)
+# TODO: can this be removed, I wish meson copied .pyx source to the build dir automatically
+# The reason we can't build the pyx files inplace and copy to build dir is because
+# the generated pxi files cannot be written to the source directory.
+# (Meson only supports out of tree builds)
+cython_sources_list = [
+    # List of cython sources e.g. .pyx, .pxd & __init__.py
+    # Does NOT include .pxi.in
+    '__init__.py',
+    'algos.pxd',
+    'algos.pyx',
+    'arrays.pxd',
+    'dtypes.pxd',
+    'hashtable.pxd',
+    'hashtable.pyx',
+    'index.pyx',
+    'indexing.pyx',
+    'internals.pyx',
+    'interval.pyx',
+    'join.pyx',
+    'khash.pxd',
+    'lib.pxd',
+    'missing.pxd',
+    'parsers.pyx',
+    'sparse.pyx',
+    'testing.pyx',
+    'tslib.pyx',
+    'util.pxd',
+]
+cython_sources = {}
+
+foreach source: cython_sources_list
+    source_pyx = fs.copyfile(source)
+    cython_sources += {source: source_pyx}
+endforeach
+
+subdir('tslibs')
+
+libs_sources = {
+    # Dict of extension name -> dict of {sources, include_dirs, and deps}
+    # numpy include dir is implicitly included
+    'algos': {'sources': [cython_sources['algos.pyx'], _algos_common_helper, _algos_take_helper, _khash_primitive_helper],
+              'include_dirs': klib_include},
```

Comment:
I think should add the generated .pxi targets as a depends argument in these calls

Reply:
They are listed as sources, so there is no need. I am going to move them to being dependencies as soon as the meson bug for this(mesonbuild/meson#9049) is addressed.

Reply:
Not sure how meson in particular works but AFAIK sources and dependencies aren't the same thing. The former is passed directly to the compiler so it knows how to build its target. The latter is used by the build system to determine the order of operations / parallelism that can be achieved. So just having it as a source doesn't prevent an issue where meson may decide to try and build say algos.c before the algos_common_helper.pxi.in target has been built

Reply:
Sources are automatically dependencies.
(xref https://ninja-build.org/manual.html#ref_dependencies)

> So just having it as a source doesn't prevent an issue where meson may decide to try and build say algos.c before the algos_common_helper.pxi.in target has been built

This is similar bug mentioned earlier that I need to patch in meson. The ninja.build file generated is unfortunately not propogating the algos_common_helper.pxi target to the target that does algos.pyx -> algos.c.

Reply:
Ah cool - thanks for the info